xclim/sdba/properties.py

```diff
@@ -947,6 +1004,138 @@ def _spatial_correlogram(da: xr.DataArray, *, dims=None, bins=100, group="time")
 )
 
 
+def _decorrelation_length(
+    da: xr.DataArray, *, radius=300, thresh=0.50, dims=None, bins=100, group="time"
+):
+    """Decorrelation length.
+
+    Distance from a grid cell where the correlation with its neighbours goes below the threshold.
+    A correlogram is calculated for each grid cell following the method from ``xclim.sdba.properties.spatial_correlogram``.
+    Then, we find the first bin closest to the correlation threshold.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+      Data.
+    radius: int
+        Radius (in km) defining the region where correlations will be calculate between a point and its neighbours.
+    thresh: float
+      Threshold correlation defining decorrelation.
+       The decorrelation length is defined as the center of the distance bin that has a correlation closest to this threshold.
+    dims: sequence of strings
+      Name of the spatial dimensions. Once these are stacked, the longitude and latitude coordinates must be 1D.
+    bins:
+      Same as argument `bins` from :py:meth:`scipy.stats.binned_statistic`.
+      If given as a scalar, the equal-width bin limits from 0 to radius are generated here
+      (instead of letting scipy do it) to improve performance.
+    group: str
+      Useless for now.
+
+    Returns
+    -------
+    xr.DataArray, [dimensionless]
+      Decorrelation length.
+
+    Notes
+    -----
+    Calculating this property requires a lot of memory. It will not work with large datasets.
+    """
+    if dims is None:
+        dims = [d for d in da.dims if d != group.dim]
+
+    corr = _pairwise_spearman(da, dims)
+
+    dists, mn, mx = _pairwise_haversine_and_bins(
+        corr.cf["longitude"].values, corr.cf["latitude"].values, transpose=True
+    )
+
+    # fill bottom triangle
+    np.nan_to_num(dists, 0)
+    dists = xr.DataArray(dists, dims=corr.dims, coords=corr.coords, name="distance")
+    trans_dists = xr.DataArray(
+        dists.T, dims=corr.dims, coords=corr.coords, name="distance"
+    )
+
+    if np.isscalar(bins):
+        bins = np.linspace(0, radius, bins + 1)
+
+    if uses_dask(corr):
+        dists = dists.chunk()
+        trans_dists = trans_dists.chunk()
+
+    w = np.diff(bins)
+    centers = xr.DataArray(
+        bins[:-1] + w / 2,
+        dims=("distance_bins",),
+        attrs={
+            "units": "km",
+            "long_name": f"Centers of the intersite distance bins (width of {w[0]:.3f} km)",
+        },
+    )
+    ds = xr.Dataset({"corr": corr, "distance": dists, "distance2": trans_dists})
+
+    # only keep points inside the radius
+    ds = ds.where(ds.distance < radius)
+
+    ds = ds.where(ds.distance2 < radius)
+
+    def _bin_corr(corr, distance):
+        """Bin and mean."""
+        return stats.binned_statistic(
+            distance, corr, statistic="mean", bins=bins
+        ).statistic
```

Comment:
I just tried the function and it works well, and faster than I expected! `binned_statistic` is a good find! Adapt my `spatial_correlogram` to use this, on the Haudenosaunee region, I went from 1m25 to 25s!

There is an issue with it however: it doesn't perform the nan-aware statistic. Points where `corr` is NaN will yield NaN bins. I think the issue gets hidden here with the use of `idxmin`, so you skip those bins already. However, in `spatial_correlogram`, it is obvious. In my temporary tests I did:
```python3
dists = dists.where(corr.notnull())
```
I'm not sure if this could be more efficient if done in numpy within `_bin_corr`.

Reply:
oh you're right, I missed that!